pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,378 @@
+	if err != nil {
+		return fmt.Errorf("Failed to create tigera-ippool-controller: %w", err)
+	}
+
+	// Watch for changes to primary resource Installation
+	err = c.WatchObject(&operator.Installation{}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return fmt.Errorf("tigera-ippool-controller failed to watch primary resource: %w", err)
+	}
+
+	// Watch for changes to APIServer
+	err = c.WatchObject(&operator.APIServer{}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		log.V(5).Info("Failed to create APIServer watch", "err", err)
+		return fmt.Errorf("apiserver-controller failed to watch primary resource: %v", err)
+	}
+
+	// Watch for changes to TigeraStatus.
+	if err = utils.AddTigeraStatusWatch(c, tigeraStatusName); err != nil {
+		return fmt.Errorf("tigera-ippool-controller failed to watch calico Tigerastatus: %w", err)
+	}
+
+	// Watch for changes to IPPool.
+	err = c.WatchObject(&crdv1.IPPool{}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return fmt.Errorf("tigera-ippool-controller failed to watch IPPool resource: %w", err)
+	}
+
+	if r.autoDetectedProvider == operator.ProviderOpenShift {
+		// Watch for openshift network configuration as well. If we're running in OpenShift, we need to
+		// merge this configuration with our own and the write back the status object.
+		err = c.WatchObject(&configv1.Network{}, &handler.EnqueueRequestForObject{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("tigera-installation-controller failed to watch openshift network config: %w", err)
+			}
+		}
+	}
+
+	// Perform periodic reconciliation. This acts as a backstop to catch reconcile issues,
+	// and also makes sure we spot when things change that might not trigger a reconciliation.
+	if err = utils.AddPeriodicReconcile(c, utils.PeriodicReconcileTime, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("tigera-ippool-controller failed to create periodic reconcile watch: %w", err)
+	}
+	return nil
+}
+
+var _ reconcile.Reconciler = &Reconciler{}
+
+type Reconciler struct {
+	config               *rest.Config
+	client               client.Client
+	scheme               *runtime.Scheme
+	watches              map[runtime.Object]struct{}
+	autoDetectedProvider operator.Provider
+	status               status.StatusManager
+}
+
+const (
+	// This label is used to track which IP pools are managed by this controller. Any IP pool
+	// with this label key/value pair is assumed to be solely managed and reconciled by this controller.
+	// We can't use OwnerReferences until this issue is resolved: https://github.com/projectcalico/calico/issues/5715
+	managedByLabel = "app.kubernetes.io/managed-by"
+	managedByValue = "tigera-operator"
+)
+
+// hasOwner returns true if the given IP pool is owned by the given Installation object, and
+// false otherwise.
+func hasOwner(pool *crdv1.IPPool, installation *operator.Installation) bool {
```

Comment:
Should we change this to `hasOwnerLabel` to make sure no one gets confused and looks for an OwnerReference?

Reply:
Ideally, this gets switched to using an OwnerReference very soon. I think it's fine to keep the name of this function generic so that the implementation can evolve without constant renaming.

Most likely, once my API bug fix goes in to enable it, we will need this function to check both the label and the owner reference for a small period of time.